src/repart/repart.c

```diff
@@ -9443,6 +9444,34 @@ static int find_root(Context *context) {
          * here. */
         r = readlink_malloc("/run/systemd/volatile-root", &device);
         if (r == -ENOENT) { /* volatile-root not found */
+                /* In case we are in the initrd, then see if we can find the actual device via the booted
+                 * ESP, since the root directory of the initrd is not what we want */
+                if (in_initrd()) {
+                        sd_id128_t loader_partition_uuid = SD_ID128_NULL;
+
+                        r = efi_loader_get_device_part_uuid(&loader_partition_uuid);
+                        if (r < 0 && r != -ENOENT && !ERRNO_IS_NEG_NOT_SUPPORTED(r))
+                                return log_error_errno(r, "Failed to determine EFI loader partition UUID: %m");
+                        if (r >= 0) {
+                                _cleanup_free_ char *esp = path_join("/dev/disk/by-partuuid/", SD_ID128_TO_UUID_STRING(loader_partition_uuid));
+                                if (!esp)
+                                        return log_oom();
+
+                                r = chase(esp, /* root= */ NULL, /* flags= */ 0, &device, /* ret_fd= */ NULL);
```

Comment:
Is this necessary? `acquire_root_devno()` internally calls `chase_and_open()`.
On failure, I think it is better to log by-partuuid path, rather than the chased path. Otherwise it is hard to understand where the device node comes from.